pkg/ingester/transfer.go

```diff
@@ -36,62 +36,74 @@ func (i *Ingester) TransferChunks(stream client.Ingester_TransferChunksServer) e
 
-		for {
-			wireSeries, err := stream.Recv()
-			if err == io.EOF {
-				break
-			}
-			if err != nil {
-				return errors.Wrap(err, "TransferChunks: Recv")
-			}
+		// run as an anonymous function so only a single error needs handling if returned
+		err := func() error {
+			for {
+				wireSeries, err := stream.Recv()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					return errors.Wrap(err, "TransferChunks: Recv")
+				}
 
-			// We can't send "extra" fields with a streaming call, so we repeat
-			// wireSeries.FromIngesterId and assume it is the same every time
-			// round this loop.
-			if fromIngesterID == "" {
-				fromIngesterID = wireSeries.FromIngesterId
-				level.Info(util.Logger).Log("msg", "processing TransferChunks request", "from_ingester", fromIngesterID)
+				// We can't send "extra" fields with a streaming call, so we repeat
+				// wireSeries.FromIngesterId and assume it is the same every time
+				// round this loop.
+				if fromIngesterID == "" {
+					fromIngesterID = wireSeries.FromIngesterId
+					level.Info(util.Logger).Log("msg", "processing TransferChunks request", "from_ingester", fromIngesterID)
+
+					// Before transfer, make sure 'from' ingester is in correct state to call ClaimTokensFor later
+					err := i.checkFromIngesterIsInLeavingState(stream.Context(), fromIngesterID)
+					if err != nil {
+						return errors.Wrap(err, "TransferChunks: checkFromIngesterIsInLeavingState")
+					}
+				}
+				descs, err := fromWireChunks(wireSeries.Chunks)
+				if err != nil {
+					return errors.Wrap(err, "TransferChunks: fromWireChunks")
+				}
 
-				// Before transfer, make sure 'from' ingester is in correct state to call ClaimTokensFor later
-				err := i.checkFromIngesterIsInLeavingState(stream.Context(), fromIngesterID)
+				state, fp, series, err := userStates.getOrCreateSeries(stream.Context(), wireSeries.UserId, wireSeries.Labels, nil)
 				if err != nil {
-					return errors.Wrap(err, "TransferChunks: checkFromIngesterIsInLeavingState")
+					return errors.Wrapf(err, "TransferChunks: getOrCreateSeries: user %s series %s", wireSeries.UserId, wireSeries.Labels)
 				}
-			}
-			descs, err := fromWireChunks(wireSeries.Chunks)
-			if err != nil {
-				return errors.Wrap(err, "TransferChunks: fromWireChunks")
-			}
+				prevNumChunks := len(series.chunkDescs)
 
-			state, fp, series, err := userStates.getOrCreateSeries(stream.Context(), wireSeries.UserId, wireSeries.Labels, nil)
-			if err != nil {
-				return errors.Wrapf(err, "TransferChunks: getOrCreateSeries: user %s series %s", wireSeries.UserId, wireSeries.Labels)
+				err = series.setChunks(descs)
+				state.fpLocker.Unlock(fp) // acquired in getOrCreateSeries
+				if err != nil {
+					return errors.Wrapf(err, "TransferChunks: setChunks: user %s series %s", wireSeries.UserId, wireSeries.Labels)
+				}
+
+				seriesReceived++
+				i.metrics.memoryChunks.Add(float64(len(series.chunkDescs) - prevNumChunks))
```

Comment:
This metric needs to be reset too.

Reply:
I don't think so, since the chunks were technically created. We just removed them. This is the same reason we don't reset `cortex_ingester_memory_series_created_total`. Instead, we just increment `cortex_ingester_memory_series_removed_total`.

Reply:
`memoryChunks` is a Gauge that tells active chunks in memory and there is no `_removed_` metric for chunks. So I guess it needs to be reset.

Reply:
I'm also convinced `memoryChunks` should be decremented.

Reply:
I'll update it to decrement if the transfer fails